### PR TITLE
sentence-case

### PR DESCRIPTION
**Pull request by retorquere**
Simpler sentence-caser. I've tried my best to make sure it is in line with the coding style for Zotero, but utilities does not have an eslint config.

### CODE REVIEW

**Comment by dstillman** [comment]
Can we add a test to https://github.com/zotero/utilities/blob/master/test/tests/utilitiesTest.js? Assuming you have the strings separately, we can just put a JSON file with the pairs in https://github.com/zotero/utilities/tree/master/test/data and load with `loadSampleData()`.

Once you've run `npm i` once, you can run tests in this repo with `npm test`.

**Comment by retorquere** [comment]
According to the README, `npm i && npm test` should run the tests, but I get 

> no such file or directory, open 'test/../resource/schema/global/schema.json'

**Comment by retorquere** [comment]
I've dumped my strings at https://gist.github.com/retorquere/8fb5a14a0b0f0a60db3df5313a258d5c . I haven't inspected everyone I must admit, this is accrued testing over the years based on user reports. I'd be happy to add tests.

**Comment by dstillman** [comment]
You probably didn't do a recursive clone. (Almost all Zotero repos require recursive clones.)

You can use `git submodule update --init --recursive` to initialize the submodules.

**Comment by retorquere** [comment]
great, thanks. But when we're talking about loadSampleData, shouldn't I be adding them to `test/tests/utilities_itemTest.js`?

**Comment by retorquere** [comment]
tests have been added.

File: utilities.js [hunk 1]
```diff
@@ -118,6 +118,76 @@ var Utilities = {
...
+		// protect nocase
+		text.replace(/<span class="nocase">.*?<\/span>|<nc>.*?<\/nc>/gi, (match, i) => {
+			preserve.push({ start: i, end: i + match.length, description: 'nocase' });
+		});
+
+		// mask html tags with characters so the sentence-casing can deal with them as simple words
+		let masked = text.replace(/<[^>]+>/g, (match, i) => {
+			preserve.push({ start: i, end: i + match.length, description: 'markup' });
+			return '\uFFFD'.repeat(match.length);
+		});
```
Comment (dstillman) [inline utilities.js:158]: The above two sections don't seem to be doing anything, at least with the sample input. Do we need them? If so, we should add some test data for them.

Comment (retorquere) [reply utilities.js:158]: The (sub-)sentence-start ones? One of them did fail a test when I removed them, but I've added two new samples. But in the mail I received you seemed to be pointing towards the (sub-)sentence start handlers, here on the site it looks like you're referring to the markup handler. I'll look into the markup handler tonight -- certainly *that* should hit on something.

Comment (dstillman) [reply utilities.js:158]: The two blocks above my comment here — `protect nocase` and `mask html tags with characters…`. Nothing failed when I removed those (but still setting `masked` properly).

Comment (retorquere) [reply utilities.js:158]: Well that's just weird. I'll look into it tonight - I have a 6 hour drive in front of me right now.

Comment (retorquere) [reply utilities.js:158]: It captures in-word markup. I will grant that I do not have a non-synthetic case at hand, but I've added one.

**Comment by dstillman** [comment]
What about the nocase?

**Comment by retorquere** [comment]
I've added a synthetic test for that.

**Comment by dstillman** [comment]
>   `"How to Get an <span class=\"nocase\">A</span> Without Trying": "How to get an <span class=\"nocase\">A</span> without trying",`

>   `"Effects of Open- and Closed-System Temperature Changes on Blood <span class=\"nocase\">O</span>₂-Binding Characteristics of <span class=\"nocase\">Atlantic</span> Bluefin Tuna (<span class=\"nocase\"><i>Thunnus thynnus</i></span>)": "Effects of open- and closed-system temperature changes on blood <span class=\"nocase\">O</span>₂-binding characteristics of <span class=\"nocase\">Atlantic</span> bluefin tuna (<span class=\"nocase\"><i>Thunnus thynnus</i></span>)",`

I don't think either of these is really appropriate as an example. CSL styles don't sentence-case, so despite the name, the sole point of `nocase` is to avoid automatic *title*-casing of foreign words, scientific terms, and mixed-case words like iPhone (though citeproc-js should arguably just avoid capitalizing the last one). So other than `Thunnus thynnus`, which _maybe_ would be protected before sentence-casing, none of the things would be. I know we're only testing the sentence-caser here, but I don't think we want to imply anywhere (even to developers looking at the code) that someone should be putting `nocase` around `A`, `Atlantic`, or `O`. Short of perhaps AI, there's no way to avoid manually recapitalizing proper nouns as a human being after sentence-casing a string, and that's not going to change with this better sentence-caser.

**Comment by retorquere** [comment]
I'd have no issue with Zotero making that call.  For BBT it's a live case, but for Zotero it doesn't need to be.

**Comment by dstillman** [comment]
I'm just suggesting we go with something like this:

> `"Migration and the Origins of <span class=\"nocase\">Homo sapiens</span>": "Migration and the origins of <span class=\"nocase\">Homo sapiens</span>",`

Which better reflects how `nocase` is meant to be used.

**Comment by dstillman** [comment]
(Which actually makes me wonder if citeproc-js actually looks for `<span class="nocase">` or just `class="nocase"`. I.e., does `<i class="nocase">` work?)

**Comment by retorquere** [comment]
> I'm just suggesting we go with something like this:
> 
> > `"Migration and the Origins of <i><span class=\"nocase\">Homo sapiens</span></i>": "Migration and the origins of <i><span class=\"nocase\">Homo sapiens</span></i>",`
> 
> Which better reflects how `nocase` is meant to be used.

Ah I see, I hadn't scrutinized the sample, this is how I got it from a user.

**Comment by retorquere** [comment]
> (Which actually makes me wonder if citeproc-js actually looks for `<span class="nocase">` or just `class="nocase"`. I.e., does `<i class="nocase">` work?)

It wouldn't be hard to add that; the BBT html-to-latex supports it but I haven't documented that. Do you know what markup citeproc supports beyond b/i/sup/sub BTW?

**Comment by retorquere** [comment]
I've replaced the nocase tests with the proposed sample

**Comment by northword** [comment]
> Maybe some of these idea can carry over? But I'd suggest e.g. testing for \p{Lu} rather than A-Z, and I think string.charAt would have problems with multibyte characters. I've opened a new PR at https://github.com/zotero/utilities/pull/26.

It looks great!

> I've tried adding chem elements detection to the sentencecaser, but it flags "No" as an element instead of as the adverb/adjective/noun. edit: also: "Be", "Au" (French), "Na" (Portuguese).

oh! Sorry, I hadn't thought about this at all. Perhaps we can apply chemical elements only when the title is in English? Or unmatch chemical elements that are function words in other languages.
This seems to be a bit of a hassle, maybe it can be done with a plugin patch as well.

**Comment by retorquere** [comment]
I think the best option would be to add an option to apply a span-nocase to the selected part of the text.

**Comment by retorquere** [comment]
quote-protection has been removed.

**Comment by dstillman** [comment]
Great! Thank you!

**Comment by retorquere** [comment]
This will show up in the next beta I take it? If so, when could we expect the next beta?

**Comment by dstillman** [comment]
Beta 32, out now

**Comment by northword** [comment]
The problem is that when the text is in all caps, the result is still in all caps, e.g. "NITROUS-OXIDE EMISSIONS FROM VEHICLES", which obviously doesn't have a specific word that needs to be capitalized, so I think that we can convert every word to lowercase when uppercase == uppercase with lowercase in all text, lowercase the each word.

**Comment by dstillman** [comment]
Oh, yes, that would be a fairly significant regression.

**Comment by retorquere** [comment]
https://github.com/zotero/utilities/pull/27

**Pull request merged**